### ISSUE TITLE
FeatureGroup ref unattainable

### ISSUE THREAD

**Issue by mbritton** [PaulLeCam/react-leaflet#210]
My FeatureGroup component has a ref attribute, and so does my Map. In componentDidMount,  I can see both objects in the console. However, when I try to view my FeatureGroup's object in the console, I get undefined.

I'm trying to use FeatureGroup's bounds to set the map's bounds. It seems strange that both refs are visible, but only one can be accessed.

```

import React, { Component } from 'react';
import { ReactDOM, render } from 'react-dom';
import { MapComponent, Map, Marker, Popup, TileLayer, GeoJson, LayerGroup, FeatureGroup } from 'react-leaflet';
import { fetchCountyData } from '../actions/countyActions.js'
import { fetchRoutes } from '../actions/routesActions.js'
import { fetchTrailheadMarkers } from '../actions/sctActions.js'
import { connect } from "react-redux"

@connect((store) => {
  return {
    countyData: store.county.countyData,
    routes: store.routes.routes,
    markers: store.markers.markers
  };
})
export default class LeafletMap extends Component {
  constructor() {
    super()

    this.state = {
      map:null,
      geoJson:null,
      bounds:null,
      countyObj:null,
      countyData:null,
      lat: 33.9301310, 
      lng: -84.7667300,
      tileURL:'http://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png',
      zoom: 10,
      markers:null,
      routes:null
    }
  }

  componentDidMount() {
    console.log('LeafletMap::componentDidMount'
    
    console.dir(this.refs) // Two sub-objects: atlMap and fg
    console.dir(this.refs['atlMap']) // Map
    console.dir(this.refs['fg']) // undefined
  }

  componentWillMount() {

  render() {
    console.log('------------------LeafletMap::render------------------')

    const position = [this.state.lat, this.state.lng]
    return (
      <Map ref="atlMap" onload={this.onMapLoaded} center={position} zoom={this.state.zoom} className="atlbike-map">
        <TileLayer
          url={this.state.tileURL} subdomains='abcd'
        />
        <FeatureGroup ref="fg">
          <GeoJson onEachFeature={this.onEachCountyFeature} data={this.props.countyData}></GeoJson>
        </FeatureGroup>
      </Map>
    )
  }
}

```

**Comment by YUzhva** [comment]
1) In my practice, I had one issue with states, when I hardcoded them in constructor() {this.state ={}} they was corrupted, so I'm always using this.setState from componentDidMount to setStates in react.

2) try to access those elements from componentDidUpdate cycle.

It looks like react need one more rendering of element to access that params.

```
componentDidMount() {
   this.setState({
      map:null,
      geoJson:null,
      bounds:null,
      countyObj:null,
      countyData:null,
      lat: 33.9301310, 
      lng: -84.7667300,
      tileURL:'http://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png',
      zoom: 10,
      markers:null,
      routes:null
    });
}

componentDidUpdate() {
   // log status to console and try to access to the elements that you need.
}
```

**Comment by PaulLeCam** [comment]
Hi,
Please make sure to read the documentation, especially the [how it works](https://github.com/PaulLeCam/react-leaflet/blob/master/docs/How%20it%20works.md) section that seem relevant to your issue.

**Comment by mbritton** [comment]
Read the documentation , and have been unable to access the ref of a geojson object despite it being defined at runtime. That is why I posted this issue.

**Comment by PaulLeCam** [comment]
Sorry if it's not explicit enough in the docs but [this part about the lifecycle](https://github.com/PaulLeCam/react-leaflet/blob/master/docs/How%20it%20works.md#lifecycle-process) means that when your component's `componentDidMount()` handler gets called, only the containing `<div>` for the map is rendered and the Leaflet map is just being created.

> 1. The top-level `Map` renders an empty `<div>` to contain the map.
> 1. The `Map`'s `componentDidMount()` handler instantiates a `Leaflet.Map()` for the created `<div>` with the component properties and sets this instance in state. This instance is made available in the context.
> 1. The `Map`'s `render()` handler is executed again, this time rendering its children components.

**Comment by samjbmason** [comment]
Just following up on this, how is it possible to access the GeoJSON components ref then if it's not available on componentDidMount ?

**Comment by scuerda** [comment]
I "solved" this with some really ugly counting in componentDidMount.

When I passed in data to my map container component, I recorded the number of geojson features that were to be added.

In comonpentDidMount, I added an event handler on the Map component for 'layeradd' and then counted how many polygons had been added and checked against what was expected. Once all were added, I had to setTimeout and then I was able to access GeoJSON feature methods and events within the setTimeout callback.

Ugly stuff, but it got the job done.

```
    componentDidMount() {
        let count = 0;
        const map = this.refs.map.leafletElement;
        map.on('layeradd', () => {
            count += 1;
            if (count > this.state.polygonCount) {
                setTimeout(()=>{this.zoomHome()}, 200)
            }
        });
    }
```
